File: server/services/zoho.js
const axios = require('axios');

let cachedToken = null;
let tokenExpiry = 0;

function getZohoDomain() {
  const tld = (process.env.ZOHO_DOMAIN || 'com').replace(/^\./, '');
  return tld;
}

async function getAccessToken() {
  if (cachedToken && Date.now() < tokenExpiry - 60000) {
    return cachedToken;
  }

  const { ZOHO_CLIENT_ID, ZOHO_CLIENT_SECRET, ZOHO_REFRESH_TOKEN } = process.env;

  if (!ZOHO_CLIENT_ID || !ZOHO_CLIENT_SECRET || !ZOHO_REFRESH_TOKEN) {
    throw new Error('Zoho credentials not configured in .env');
  }

  const tld = getZohoDomain();

  const params = new URLSearchParams({
    refresh_token: ZOHO_REFRESH_TOKEN,
    client_id: ZOHO_CLIENT_ID,
    client_secret: ZOHO_CLIENT_SECRET,
    grant_type: 'refresh_token',
  });

  const response = await axios.post(
    `https://accounts.zoho.${tld}/oauth/v2/token`,
    params.toString(),
    { headers: { 'Content-Type': 'application/x-www-form-urlencoded' } }
  );

  if (!response.data.access_token) {
    throw new Error(`Zoho token error: ${JSON.stringify(response.data)}`);
  }

  cachedToken = response.data.access_token;
  tokenExpiry = Date.now() + (response.data.expires_in || 3600) * 1000;
  return cachedToken;
}

function zohoError(context, err) {
  if (err.response) {
    const body = err.response.data;
    const msg = typeof body === 'object' ? JSON.stringify(body) : String(body);
    return new Error(`Zoho ${context} [${err.response.status}]: ${msg || err.message}`);
  }
  return err;
}

async function fetchAccounts(orgId) {
  const token = await getAccessToken();
  const tld = getZohoDomain();
  try {
    const response = await axios.get(
      `https://www.zohoapis.${tld}/books/v3/chartofaccounts?organization_id=${orgId}`,
      {
        headers: { Authorization: `Zoho-oauthtoken ${token}` },
        params: { per_page: 200 },
      }
    );

    const data = response.data;
    if (data.code !== 0) {
      throw new Error(`Zoho accounts error [${data.code}]: ${data.message}`);
    }

    return (data.chartofaccounts || []).map((a) => ({
      id: a.account_id,
      name: a.account_name,
      type: a.account_type,
    }));
  } catch (err) {
    throw zohoError('accounts', err);
  }
}

async function postJournalEntry(orgId, payload) {
  const token = await getAccessToken();
  const tld = getZohoDomain();
  const url = `https://www.zohoapis.${tld}/books/v3/manualjournals`;
  console.log('[Zoho POST]', url, 'org:', orgId);
  console.log('[Zoho payload]', JSON.stringify(payload));
  try {
    const response = await axios.post(
      url,
      payload,
      {
        headers: {
          Authorization: `Zoho-oauthtoken ${token}`,
          'Content-Type': 'application/json',
        },
        params: { organization_id: String(orgId).trim() },
      }
    );

    const data = response.data;
    console.log('[Zoho response]', JSON.stringify(data));
    if (data.code !== 0) {
      throw new Error(`Zoho JE error [${data.code}]: ${data.message}`);
    }

    return data.journal;
  } catch (err) {
    if (err.response) {
      console.error('[Zoho error]', err.response.status, JSON.stringify(err.response.data));
    }
    throw zohoError('journal entry', err);
  }
}

module.exports = { getAccessToken, fetchAccounts, postJournalEntry };

File: app/services/zoho.py
import time
import httpx
from app.config import ZOHO_CLIENT_ID, ZOHO_CLIENT_SECRET, ZOHO_REFRESH_TOKEN, ZOHO_DOMAIN

_cached_token: str = ""
_token_expiry: float = 0.0


async def get_access_token() -> str:
    global _cached_token, _token_expiry
    if _cached_token and time.time() < _token_expiry - 60:
        return _cached_token

    if not ZOHO_CLIENT_ID or not ZOHO_CLIENT_SECRET or not ZOHO_REFRESH_TOKEN:
        raise RuntimeError("Zoho credentials not configured")

    async with httpx.AsyncClient() as client:
        resp = await client.post(
            f"https://accounts.zoho.{ZOHO_DOMAIN}/oauth/v2/token",
            data={
                "refresh_token": ZOHO_REFRESH_TOKEN,
                "client_id": ZOHO_CLIENT_ID,
                "client_secret": ZOHO_CLIENT_SECRET,
                "grant_type": "refresh_token",
            },
        )
        data = resp.json()

    if "access_token" not in data:
        raise RuntimeError(f"Zoho token error: {data}")

    _cached_token = data["access_token"]
    _token_expiry = time.time() + int(data.get("expires_in", 3600))
    return _cached_token


def _zoho_base() -> str:
    return f"https://www.zohoapis.{ZOHO_DOMAIN}/books/v3"


async def fetch_accounts(org_id: str) -> list[dict]:
    token = await get_access_token()
    accounts = []
    page = 1
    async with httpx.AsyncClient(timeout=30) as client:
        while True:
            resp = await client.get(
                f"{_zoho_base()}/chartofaccounts",
                headers={"Authorization": f"Zoho-oauthtoken {token}"},
                params={"organization_id": org_id, "per_page": 200, "page": page},
            )
            data = resp.json()
            if data.get("code") != 0:
                raise RuntimeError(f"Zoho accounts error [{data.get('code')}]: {data.get('message')}")
            for a in data.get("chartofaccounts", []):
                accounts.append({"id": a["account_id"], "name": a["account_name"], "type": a["account_type"]})
            if not data.get("page_context", {}).get("has_more_page"):
                break
            page += 1
    return accounts


async def post_journal_entry(org_id: str, payload: dict) -> dict:
    token = await get_access_token()
    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.post(
            f"{_zoho_base()}/journals",
            headers={"Authorization": f"Zoho-oauthtoken {token}", "Content-Type": "application/json"},
            params={"organization_id": str(org_id).strip()},
            json=payload,
        )
        data = resp.json()
    if data.get("code") != 0:
        raise RuntimeError(f"Zoho JE error [{data.get('code')}]: {data.get('message')}")
    return data.get("journal", {})


async def create_expense(org_id: str, payload: dict) -> dict:
    token = await get_access_token()
    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.post(
            f"{_zoho_base()}/expenses",
            headers={"Authorization": f"Zoho-oauthtoken {token}", "Content-Type": "application/json"},
            params={"organization_id": str(org_id).strip()},
            json=payload,
        )
        data = resp.json()
    if data.get("code") != 0:
        raise RuntimeError(f"Zoho Expense error [{data.get('code')}]: {data.get('message')}")
    return data.get("expense", {})


async def attach_journal_document(org_id: str, journal_id: str, file_bytes: bytes, filename: str, mime_type: str) -> dict:
    token = await get_access_token()
    async with httpx.AsyncClient(timeout=60) as client:
        resp = await client.post(
            f"{_zoho_base()}/journals/{journal_id}/documents",
            headers={"Authorization": f"Zoho-oauthtoken {token}"},
            params={"organization_id": str(org_id).strip()},
            files={"attachment": (filename, file_bytes, mime_type)},
        )
        data = resp.json()
    if data.get("code") != 0:
        raise RuntimeError(f"Zoho attach error [{data.get('code')}]: {data.get('message')}")
    return data


async def attach_expense_document(org_id: str, expense_id: str, file_bytes: bytes, filename: str, mime_type: str) -> dict:
    token = await get_access_token()
    async with httpx.AsyncClient(timeout=60) as client:
        resp = await client.post(
            f"{_zoho_base()}/expenses/{expense_id}/receipts",
            headers={"Authorization": f"Zoho-oauthtoken {token}"},
            params={"organization_id": str(org_id).strip()},
            files={"attachment": (filename, file_bytes, mime_type)},
        )
        data = resp.json()
    if data.get("code") != 0:
        raise RuntimeError(f"Zoho expense attach error [{data.get('code')}]: {data.get('message')}")
    return data


async def search_contact_by_name(org_id: str, name: str) -> str | None:
    """Return contact_id if a Zoho contact matches this name exactly (case-insensitive), else None."""
    token = await get_access_token()
    async with httpx.AsyncClient(timeout=15) as client:
        resp = await client.get(
            f"{_zoho_base()}/contacts",
            headers={"Authorization": f"Zoho-oauthtoken {token}"},
            params={"organization_id": str(org_id).strip(), "search_text": name, "per_page": 10},
        )
        data = resp.json()
    for c in data.get("contacts", []):
        if c.get("contact_name", "").strip().lower() == name.strip().lower():
            return c["contact_id"]
    return None


def clear_token_cache() -> None:
    global _cached_token, _token_expiry
    _cached_token = ""
    _token_expiry = 0.0
